### ISSUE TITLE
Incorrect results after pool is compacted

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#4709]
Repro is with Zed commit d64a090. Given what I've observed, it seems likely this issue shares a root cause with some of our existing known compaction bugs #4661 and/or #4666, but since this exhibits a different symptom than previously observed I'm opening this separately and will check if it's gone should one of those others get fixed first.

Here's a repro script that loads the wrccdc data set (available at `s3://brim-sampledata/wrccdc/zeek-logs`) and runs the same time-bucketed `count()` before and after compaction. This is a query from the Rug tests.

```
rm -f pre-compact post-compact
zed create -use wrccdc
zed load ~/Downloads/wrccdc/[!l]*.log.gz
zed query -Z 'from wrccdc | count() by every(1h) | sort -r -nulls last ts' > pre-compact
zed manage -once
zed query -Z 'from wrccdc | count() by every(1h) | sort -r -nulls last ts' > post-compact
diff pre-compact post-compact
```

I've attached [results.tgz](https://github.com/brimdata/zed/files/12017562/results.tgz) that has those full final outputs, but here's the important part: The bucket for key `ts: 2018-03-24T22:00:00Z` appears _twice_ in the post-compaction query output.

```
$ diff -y pre-compact post-compact 
...
{									{
    ts: 2018-03-24T22:00:00Z,						    ts: 2018-03-24T22:00:00Z,
    count: 12144970 (uint64)					|           count: 11304325 (uint64)
								>       }
								>       {
								>           ts: 2018-03-24T22:00:00Z,
								>           count: 840645 (uint64)
}									}
...
```

Since I know this data set well and am familiar with the bugs we already have open in this area, I immediately suspected the records from `smb_mapping.log.gz` were triggering the problem, since this includes some values that become null pool keys and we've seen them cause exceptions in the past.

```
$ zq 'ts==null | count()' smb_mapping.log.gz 
26(uint64)
```

If I repeat the repro but with that particular log file excluded during the `zed load`, the query result becomes the same post-compaction as it was pre-compaction.

**Comment by philrz** [comment]
Update: I'd hoped the changes in #4723 might also address this issue, but I just went through the repro steps again at current Zed commit b566f6d and the symptoms are the same as before.

**Comment by mattnibs** [comment]
This behavior is actually not related to lake manager, but is rather exposed by lake manager putting objects in an optimized order. Here's a simple replication of the same underlying bug:

```bash
zed create -use test
echo '{ts:150} {ts:null}' | zed load -
echo '{ts:1}' | zed load -
zed query 'head 1'
```
The output of the query is `{ts:1}` when it should be `{ts:150}` since pool `test` is in descending order and nulls should be the minimum value.

**Comment by philrz** [comment]
It does indeed look like a legitimate issue was found and fixed in the linked PR #4735 and perhaps that's a prerequisite to fixing the originally-cited symptom that was found while testing compaction. However, I'm reopening this one because my attempt to verify this as fixed in Zed commit b8e1b18 was unsuccessful. Repeating the original repro steps:

```
$ zed -version
Version: v1.9.0-7-gb8e1b188

$ rm -f pre-compact post-compact

$ zed create -use wrccdc
pool created: wrccdc 2TfdB4vHD8qQmekYjK0bLS8bD01
Switched to branch "main" on pool "wrccdc"

$ zed load ~/Downloads/wrccdc/[!l]*.log.gz
(34/34) 17.74GB/17.74GB 8.33MB/s 100.00%

$ zed query -Z 'from wrccdc | count() by every(1h) | sort -r -nulls last ts' > pre-compact

$ zed manage
{"level":"info","ts":1691439398.00351,"logger":"pool","msg":"updating pool","name":"wrccdc","id":"2TfdB4vHD8qQmekYjK0bLS8bD01","branch":"main","interval":60}
{"level":"info","ts":1691439904.304228,"logger":"pool","msg":"compaction completed","name":"wrccdc","id":"2TfdB4vHD8qQmekYjK0bLS8bD01","branch":"main","interval":60,"runs_found":1,"objects_compacted":100}

$ zed query -Z 'from wrccdc | count() by every(1h) | sort -r -nulls last ts' > post-compact

$ diff -y pre-compact post-compact
...
{									{
    ts: 2018-03-24T22:00:00Z,						    ts: 2018-03-24T22:00:00Z,
    count: 12144970 (uint64)					|           count: 11304325 (uint64)
}									}
{									{
								>           ts: 2018-03-24T22:00:00Z,
								>           count: 840645 (uint64)
								>       }
								>       {
    ts: 2018-03-24T21:00:00Z,						    ts: 2018-03-24T21:00:00Z,
    count: 10539392 (uint64)						    count: 10539392 (uint64)
}									}
...
{									{
    ts: 2018-03-24T03:00:00Z,						    ts: 2018-03-24T03:00:00Z,
    count: 17531606 (uint64)					|           count: 5017699 (uint64)
}									}
{									{
								>           ts: 2018-03-24T03:00:00Z,
								>           count: 12513907 (uint64)
								>       }
								>       {
    ts: 2018-03-24T02:00:00Z,						    ts: 2018-03-24T02:00:00Z,
    count: 36071360 (uint64)						    count: 36071360 (uint64)
}									}
{									{
    ts: 2018-03-24T01:00:00Z,						    ts: 2018-03-24T01:00:00Z,
    count: 29723753 (uint64)						    count: 29723753 (uint64)
}									}
{									{
    ts: 2018-03-24T00:00:00Z,						    ts: 2018-03-24T00:00:00Z,
    count: 66013903 (uint64)					|           count: 54014474 (uint64)
}									}
{									{
								>           ts: 2018-03-24T00:00:00Z,
								>           count: 11999429 (uint64)
								>       }
								>       {
    ts: 2018-03-23T23:00:00Z,						    ts: 2018-03-23T23:00:00Z,
    count: 35900050 (uint64)					|           count: 22981772 (uint64)
								>       }
								>       {
								>           ts: 2018-03-23T23:00:00Z,
								>           count: 12918278 (uint64)
}									}
...
```

The `diff` output above has been trimmed to just show the parts with differences. But as you can see, there's now multiple spots where the same time bucket is output twice whereas previously it only happened once.

**Comment by philrz** [comment]
Verified in Zed commit 08fc271.

Repeating the repro steps, we now see that the query result is the same both before and after compaction.

```
$ zed -version
Version: v1.9.0-10-g08fc271c

$ rm -f pre-compact post-compact

$ zed create -use wrccdc
pool created: wrccdc 2U5aFejUX4O8hHHWtewCQIxOC6k
Switched to branch "main" on pool "wrccdc"

$ zed load ~/Downloads/wrccdc/[!l]*.log.gz
(34/34) 17.74GB/17.74GB 4.32MB/s 100.00%
2U5cfvVnRGzExPjoqBurAbVqBId committed

$ zed query -Z 'from wrccdc | count() by every(1h) | sort -r -nulls last ts' > pre-compact

$ zed manage
{"level":"info","ts":1692234330.753982,"logger":"pool","msg":"updating pool","name":"wrccdc","id":"2U5aFejUX4O8hHHWtewCQIxOC6k","branch":"main","interval":60}
{"level":"info","ts":1692234861.711176,"logger":"pool","msg":"compaction completed","name":"wrccdc","id":"2U5aFejUX4O8hHHWtewCQIxOC6k","branch":"main","interval":60,"runs_found":1,"objects_compacted":100}

$ zed query -Z 'from wrccdc | count() by every(1h) | sort -r -nulls last ts' > post-compact

$ diff -y pre-compact post-compact
[no output]
```

Thanks @mattnibs!
